### PR TITLE
Support apiserver customization EventRateLimit

### PR DESCRIPTION
**Pull request by zhouke1991**
### What this PR does / why we need it
support specifying apiserverEventRateLimitConf by APISERVER_EVENT_RATE_LIMIT_CONF_BASE64 in cluster config file

### CODE REVIEW

File: packages/tkg-clusterclass-aws/bundle/config/upstream/base.yaml [hunk 1]
```diff
@@ -1146,6 +1158,73 @@ spec:
...
+                  admission-control-config-file: "/etc/kubernetes/admission-control-config.yaml"
+            - op: add
+              path: /spec/template/spec/kubeadmConfigSpec/files
+              valueFrom:
+                template: |
+                  - path: /etc/kubernetes/admission-control-config.yaml
+                    owner: "root:root"
+                    permissions: "0600"
+                    encoding: base64
+                    content: YXBpVmVyc2lvbjogYXBpc2VydmVyLmNvbmZpZy5rOHMuaW8vdjEKa2luZDogQWRtaXNzaW9uQ29uZmlndXJhdGlvbgpwbHVnaW5zOgotIG5hbWU6IEV2ZW50UmF0ZUxpbWl0CiAgcGF0aDogZXZlbnRjb25maWcueWFtbAoK
```
Comment (DanielXiao) [inline packages/tkg-clusterclass-aws/bundle/config/upstream/base.yaml:1212]: Could you use a ytt encode function to generate base64 content? It will be easy to review or change in the future.

**Comment by randomvariable** [comment]
We've had some internal discussion on the utility of these, which is pending resolution.

File: packages/tkg-clusterclass-aws/bundle/config/upstream/base.yaml [hunk 2]
```diff
@@ -1146,6 +1158,73 @@ spec:
...
+              path: /spec/template/spec/kubeadmConfigSpec/clusterConfiguration/apiServer/extraArgs
+              valueFrom:
+                template: |
+                  enable-admission-plugins: "NodeRestriction,EventRateLimit"
+                  admission-control-config-file: "/etc/kubernetes/admission-control-config.yaml"
+            - op: add
+              path: /spec/template/spec/kubeadmConfigSpec/files
+              valueFrom:
+                template: |
+                  - path: /etc/kubernetes/admission-control-config.yaml
```
Comment (randomvariable) [inline packages/tkg-clusterclass-aws/bundle/config/upstream/base.yaml:1208]: This now conflicts with PodSecurityAdmission ...

File: packages/tkg-clusterclass-aws/bundle/config/upstream/base.yaml [hunk 3]
```diff
@@ -1146,6 +1158,73 @@ spec:
...
+  - name: apiserverEventRateLimitConf
+    definitions:
+      - selector:
+          apiVersion: controlplane.cluster.x-k8s.io/v1beta1
+          kind: KubeadmControlPlaneTemplate
+          matchResources:
+            controlPlane: true
+          jsonPatches:
+            - op: add
+              path: /spec/template/spec/kubeadmConfigSpec/clusterConfiguration/apiServer/extraArgs
```
Comment (lubronzhan) [inline packages/tkg-clusterclass-aws/bundle/config/upstream/base.yaml:1199]: Just learnt from other issue we've seen, this will overwrite the whole extraArgs.
If you check this slack https://vmware.slack.com/archives/CEVS7024C/p1670354781024059?thread_ts=1670349930.210859&cid=CEVS7024C you will see example we did

We changed from 
```
        path: /spec/template/spec/kubeadmConfigSpec/clusterConfiguration/apiServer
        valueFrom:
          template: |
            certSANs:
            {{- range .extraDNS }}
            - {{ . }}
            {{- end }}

```
To
```
        path: /spec/template/spec/kubeadmConfigSpec/clusterConfiguration/apiServer/certSANs
        valueFrom:
          template: |
            {{- range .extraDNS }}
            - {{ . }}
            {{- end }}
```

File: packages/tkg-clusterclass-aws/bundle/config/upstream/base.yaml [hunk 4]
```diff
@@ -1146,6 +1158,73 @@ spec:
...
+              path: /spec/template/spec/kubeadmConfigSpec/files
+              valueFrom:
+                template: |
+                  - path: /etc/kubernetes/encryption-config.yaml
+                    owner: "root:root"
+                    permissions: "0600"
+                    encoding: base64
+                    content: {{ $apiserverEncryptionConf }}
+            - op: add
+              path: /spec/template/spec/kubeadmConfigSpec/clusterConfiguration/apiServer/extraArgs
```
Comment (lubronzhan) [inline packages/tkg-clusterclass-aws/bundle/config/upstream/base.yaml:1179]: Same here

File: packages/tkg-clusterclass-aws/bundle/config/upstream/base.yaml [hunk 5]
```diff
@@ -1146,6 +1158,73 @@ spec:
...
+          apiVersion: controlplane.cluster.x-k8s.io/v1beta1
+          kind: KubeadmControlPlaneTemplate
+          matchResources:
+            controlPlane: true
+          jsonPatches:
+            - op: add
+              path: /spec/template/spec/kubeadmConfigSpec/files
+              valueFrom:
+                template: |
+                  - path: /etc/kubernetes/encryption-config.yaml
```
Comment (chrischdi) [inline packages/tkg-clusterclass-aws/bundle/config/upstream/base.yaml:1173]:
```suggestion
              path: /spec/template/spec/kubeadmConfigSpec/files/-
              valueFrom:
                template: |
                    path: /etc/kubernetes/encryption-config.yaml
```

```suggestion
              path: /spec/template/spec/kubeadmConfigSpec/files/-
```

This needs to append to files, to not replace already existing files.

File: packages/tkg-clusterclass-aws/bundle/config/upstream/base.yaml [hunk 6]
```diff
@@ -1146,6 +1158,73 @@ spec:
...
+            - op: add
+              path: /spec/template/spec/kubeadmConfigSpec/clusterConfiguration/apiServer/extraArgs
+              valueFrom:
+                template: |
+                  encryption-provider-config: "/etc/kubernetes/encryption-config.yaml"
+            - op: add
+              path: /spec/template/spec/kubeadmConfigSpec/clusterConfiguration/apiServer/extraVolumes
+              valueFrom:
+                template: |
+                  - name: encryption-conf
```
Comment (chrischdi) [inline packages/tkg-clusterclass-aws/bundle/config/upstream/base.yaml:1187]:
```suggestion
              path: /spec/template/spec/kubeadmConfigSpec/clusterConfiguration/apiServer/extraVolumes/-
              valueFrom:
                template: |
                    name: encryption-conf
```

same here.

File: packages/tkg-clusterclass-aws/bundle/config/upstream/base.yaml (hunk 2 above)
Comment (chrischdi) [reply packages/tkg-clusterclass-aws/bundle/config/upstream/base.yaml:1208]: We will have to combine the templating for this file to leverage both configuration features.

File: packages/tkg-clusterclass-aws/bundle/config/upstream/overlay-pod-security-admission.yaml [hunk 7]
```diff
@@ -64,71 +64,104 @@ spec:
                 items:
                   type: string
                 description: "namespaces excluded to apply PodSecurityConfiguration Admission."
-  #@overlay/match missing_ok=True
-  patches:
   #@overlay/remove
   #@overlay/match by=overlay.map_key("name"),missing_ok=True
-  - name: podSecurityStandard
+  - name: apiserverEventRateLimitConf
```
Comment (chrischdi) [inline packages/tkg-clusterclass-aws/bundle/config/upstream/overlay-pod-security-admission.yaml:69]:
```suggestion
  - name: podSecurityStandard
```

The removal stays the same, the upstream templates have a `podSecurityStandard` entry which needs to get removed if exists.

File: packages/tkg-clusterclass-aws/bundle/config/upstream/overlay-pod-security-admission.yaml [hunk 8]
```diff
@@ -64,71 +64,104 @@ spec:
...
+                template: |
+                  {{ $enableAdmissionPlugin := false }}
+                  {{- range $key, $val := .apiServerExtraArgs }}
+                  {{- if eq $key "enable-admission-plugin" }}
+                  {{- $enableAdmissionPlugin = true }}
+                  {{- end }}
+                  {{- if not $enableAdmissionPlugin }}
+                  enable-admission-plugins: "NodeRestriction,EventRateLimit"
+                  {{- end }}
+                  admission-control-config-file: "/etc/kubernetes/admission-control-config.yaml"
```
Comment (chrischdi) [inline packages/tkg-clusterclass-aws/bundle/config/upstream/overlay-pod-security-admission.yaml:104]: The args need to get separated, one OP per extraArg. Currently this would add a single extraArg called `admission-control-config-file` (see the last part of the path key) and add a multi-line string as value.

Regarding the `enable-admission-plugins`: I think this needs to get separated to an separate patch (element in `.spec.topology.patches`). I think you not always want to add the flag right? This could only be done by using the `enabled_if` in a patch.
```suggestion
            - op: add
              path: "/spec/template/spec/kubeadmConfigSpec/clusterConfiguration/apiServer/extraArgs/admission-control-config-file"
              value: "/etc/kubernetes/admission-control-config.yaml"
```

File: packages/tkg-clusterclass-aws/bundle/config/upstream/overlay-pod-security-admission.yaml [hunk 9]
```diff
@@ -64,71 +64,104 @@ spec:
...
+              valueFrom:
+                template: |
+                  - name: admin-control-conf
+                    hostPath: /etc/kubernetes/admission-control-config.yaml
+                    mountPath: /etc/kubernetes/admission-control-config.yaml
+                  {{ if .apiserverEventRateLimitConf }}
+                  - name: event-conf
+                    hostPath: /etc/kubernetes/eventconfig.yaml
+                    mountPath: /etc/kubernetes/eventconfig.yaml
+                  {{- end}}
```
Comment (chrischdi) [inline packages/tkg-clusterclass-aws/bundle/config/upstream/overlay-pod-security-admission.yaml:116]: Also needs to get separated and event-conf get added in separate patch, where `enabled_if` decides if it is to add or not
```suggestion
            - op: add
              path: "/spec/template/spec/kubeadmConfigSpec/clusterConfiguration/apiServer/extraVolumes/-"
              valueFrom:
                template: |
                  name: admin-control-conf
                  hostPath: /etc/kubernetes/admission-control-config.yaml
                  mountPath: /etc/kubernetes/admission-control-config.yaml
```

File: packages/tkg-clusterclass-aws/bundle/config/upstream/overlay-pod-security-admission.yaml [hunk 10]
```diff
@@ -64,71 +64,104 @@ spec:
...
+                          runtimeClasses: []
+                          namespaces: [{{ $namespace_exemptions }}]
+                    {{- end }}
+                    {{ if .apiserverEventRateLimitConf }}
+                    - name: EventRateLimit
+                      path: eventconfig.yaml
+                    path: /etc/kubernetes/eventconfig.yaml
+                    content: |-
+                      {{ .apiserverEventRateLimitConf}}
+                    {{- end }}
```
Comment (chrischdi) [inline packages/tkg-clusterclass-aws/bundle/config/upstream/overlay-pod-security-admission.yaml:166]: also here: needs to be a separate patch, the `add` json operation only allows to add a single entry to the array.
Also toggle it in a separate patch via `enabledIf: {{ .apiserverEventRateLimitConf}}`

or similar.

File: providers/infrastructure-aws/v2.0.2/cconly/overlay-kube-apiserver-admission.yaml [hunk 11]
```diff
@@ -0,0 +1,192 @@
...
+            description: "exemption configuration for the PodSecurityConfiguration."
+            properties:
+              namespaces:
+                type: array
+                items:
+                  type: string
+                description: "namespaces excluded to apply PodSecurityConfiguration Admission."
+  #@overlay/remove
+  #@overlay/match by=overlay.map_key("name"),missing_ok=True
+  - name: eventRateLimitConf
```
Comment (chrischdi) [inline providers/infrastructure-aws/v2.0.2/cconly/overlay-kube-apiserver-admission.yaml:69]:
```suggestion
```

This should not be required

File: providers/infrastructure-aws/v2.0.2/cconly/overlay-kube-apiserver-admission.yaml [hunk 12]
```diff
@@ -0,0 +1,192 @@
...
+    required: false
+    schema:
+      openAPIV3Schema:
+        type: string
+        default: ""
+  #@overlay/match missing_ok=True
+  patches:
+  #@overlay/remove
+  #@overlay/match by=overlay.map_key("name"),missing_ok=True
+  - name: admissionConfiguration
```
Comment (chrischdi) [inline providers/infrastructure-aws/v2.0.2/cconly/overlay-kube-apiserver-admission.yaml:81]:
```suggestion
  - name: podSecurityStandard
```
This needs to keep on removing the upstream patch for `podSecurityStandard`

File: providers/infrastructure-aws/v2.0.2/cconly/overlay-kube-apiserver-admission.yaml [hunk 13]
```diff
@@ -0,0 +1,192 @@
...
+        value: "/etc/kubernetes/admission-control-config.yaml"
+      - op: add
+        path: "/spec/template/spec/kubeadmConfigSpec/clusterConfiguration/apiServer/extraVolumes/-"
+        valueFrom:
+          template: |
+            - name: admin-control-conf
+              hostPath: /etc/kubernetes/admission-control-config.yaml
+              mountPath: /etc/kubernetes/admission-control-config.yaml
+              readOnly: true
+              pathType: "File"
```
Comment (chrischdi) [inline providers/infrastructure-aws/v2.0.2/cconly/overlay-kube-apiserver-admission.yaml:103]:
```suggestion
            name: admin-control-conf
            hostPath: /etc/kubernetes/admission-control-config.yaml
            mountPath: /etc/kubernetes/admission-control-config.yaml
            readOnly: true
            pathType: "File"
```
This needs to be a dict because we append to the array, not an array

File: providers/infrastructure-aws/v2.0.2/cconly/overlay-kube-apiserver-admission.yaml [hunk 14]
```diff
@@ -0,0 +1,192 @@
...
+              {{ $defaultWarnAudit := "baseline" }}
+              {{- if .podSecurityStandard.exemptions.namespaces -}}
+                {{ range $namespace := .podSecurityStandard.exemptions.namespaces -}}
+                  {{ $namespace_exemptions = printf "%s, %q" $namespace_exemptions $namespace -}}
+                {{- end -}}
+              {{- end -}}
+              apiVersion: apiserver.config.k8s.io/v1
+              kind: AdmissionConfiguration
+              plugins:
+              {{ if and (not .podSecurityStandard.deactivated) (semverCompare ">= v1.24" .builtin.controlPlane.version) }}
```
Comment (chrischdi) [inline providers/infrastructure-aws/v2.0.2/cconly/overlay-kube-apiserver-admission.yaml:120]:
```suggestion
              {{ if and (not .podSecurityStandard.deactivated) (semverCompare ">= v1.24" .builtin.controlPlane.version) }}
              {{ $namespace_exemptions := printf "%q, %q" "kube-system" "tkg-system" -}}
              {{ $defaultWarnAudit := "baseline" }}
              {{- if .podSecurityStandard.exemptions.namespaces -}}
                {{ range $namespace := .podSecurityStandard.exemptions.namespaces -}}
                  {{ $namespace_exemptions = printf "%s, %q" $namespace_exemptions $namespace -}}
                {{- end -}}
              {{- end -}}
```

File: providers/infrastructure-aws/v2.0.2/cconly/overlay-kube-apiserver-admission.yaml [hunk 15]
```diff
@@ -0,0 +1,192 @@
...
+          template: |-
+            path: /etc/kubernetes/admission-control-config.yaml
+            content: |-
+              {{ $namespace_exemptions := printf "%q, %q" "kube-system" "tkg-system" -}}
+              {{ $defaultWarnAudit := "baseline" }}
+              {{- if .podSecurityStandard.exemptions.namespaces -}}
+                {{ range $namespace := .podSecurityStandard.exemptions.namespaces -}}
+                  {{ $namespace_exemptions = printf "%s, %q" $namespace_exemptions $namespace -}}
+                {{- end -}}
+              {{- end -}}
```
Comment (chrischdi) [inline providers/infrastructure-aws/v2.0.2/cconly/overlay-kube-apiserver-admission.yaml:116]:
```suggestion
```
Let's move this down (see suggestion below)

File: providers/infrastructure-aws/v2.0.2/cconly/overlay-kube-apiserver-admission.yaml [hunk 16]
```diff
@@ -0,0 +1,192 @@
...
+                    namespaces: [{{ $namespace_exemptions }}]
+              {{- end }}
+              {{ if .eventRateLimitConf }}
+              - name: EventRateLimit
+                path: eventConfig.yaml
+              {{- end }}
+    enabledIf: '{{ or (and (not .podSecurityStandard.deactivated) (semverCompare ">= v1.24" .builtin.controlPlane.version)) (.eventRateLimitConf) }}'
+    #@overlay/remove
+  #@overlay/match by=overlay.map_key("name"),missing_ok=True
+  - name: eventRateLimitConf
```
Comment (chrischdi) [inline providers/infrastructure-aws/v2.0.2/cconly/overlay-kube-apiserver-admission.yaml:154]:
```suggestion
```

It does not exist, so removal not required.

File: providers/infrastructure-aws/v2.0.2/cconly/overlay-kube-apiserver-admission.yaml [hunk 17]
```diff
@@ -0,0 +1,192 @@
...
+    description: "Adds an admission configuration for EventRateLimit."
+    definitions:
+    - selector:
+        apiVersion: controlplane.cluster.x-k8s.io/v1beta1
+        kind: KubeadmControlPlaneTemplate
+        matchResources:
+          controlPlane: true
+      jsonPatches:
+      - op: add
+        path: "/spec/template/spec/kubeadmConfigSpec/clusterConfiguration/apiServer/extraArgs/-"
```
Comment (chrischdi) [inline providers/infrastructure-aws/v2.0.2/cconly/overlay-kube-apiserver-admission.yaml:166]:
```suggestion
        path: "/spec/template/spec/kubeadmConfigSpec/clusterConfiguration/apiServer/extraArgs/enable-admission-plugins"
```

Extra arg is a map, not an array. `/-` to append could only be used for arrays.

File: providers/infrastructure-aws/v2.0.2/cconly/overlay-kube-apiserver-admission.yaml [hunk 18]
```diff
@@ -0,0 +1,192 @@
...
+        valueFrom:
+          template: |
```
Comment (chrischdi) [inline providers/infrastructure-aws/v2.0.2/cconly/overlay-kube-apiserver-admission.yaml:168]: This template won't work.

1. You have to drop `enable-admission-plugins: ` in line 176
2. If the template results in empty string, it will set the argument to empty string.
  * to decide wether or not this extraArg should get updated, you should use `enabled_if`.

File: providers/infrastructure-aws/v2.0.2/cconly/overlay-kube-apiserver-admission.yaml [hunk 19]
```diff
@@ -0,0 +1,176 @@
...
+    definitions:
+    - selector:
+        apiVersion: controlplane.cluster.x-k8s.io/v1beta1
+        kind: KubeadmControlPlaneTemplate
+        matchResources:
+          controlPlane: true
+      jsonPatches:
+      - op: add
+        path: "/spec/template/spec/kubeadmConfigSpec/clusterConfiguration/apiServer/extraArgs/enable-admission-plugins"
+        value: "NodeRestriction,EventRateLimit"
```
Comment (zhouke1991) [inline providers/infrastructure-aws/v2.0.2/cconly/overlay-kube-apiserver-admission.yaml:161]: Using append

File: providers/infrastructure-vsphere/v1.5.1/cconly/base.yaml [hunk 20]
```diff
@@ -2,6 +2,7 @@
 #@ load("@ytt:data", "data")
 #@ load("@ytt:regexp", "regexp")
 
+---
```
Comment (zhouke1991) [inline providers/infrastructure-vsphere/v1.5.1/cconly/base.yaml:5]: Remove empty line

File: providers/infrastructure-oci/v0.4.0/cconly/base.yaml [hunk 21]
```diff
@@ -254,7 +254,7 @@ spec:
       schema:
         openAPIV3Schema:
           type: string
-          default: TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256,TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256,TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305,TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384,TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305,TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384
+      default: TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256,TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256,TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305,TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384,TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305,TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384
```
Comment (zhouke1991) [inline providers/infrastructure-oci/v0.4.0/cconly/base.yaml:257]: revert

File: packages/tkg-clusterclass-aws/bundle/config/upstream/overlay-kube-apiserver-admission.yaml [hunk 22]
```diff
@@ -130,5 +140,51 @@ spec:
...
+      - op: add
+        path: "/spec/template/spec/kubeadmConfigSpec/clusterConfiguration/apiServer/extraArgs/enable-admission-plugins"
+        valueFrom:
+          template: |
+            {{ $containEnableAdmissionPlugin := false }}
+            {{- $admissionPlugins := "" }}
+            {{- range $key, $val := .apiServerExtraArgs }}
+            {{- if eq $key "enable-admission-plugins" }}
+              {{- $containEnableAdmissionPlugin = true }}
+              {{- $admissionPlugins = $val }}
```
Comment (wjun) [inline packages/tkg-clusterclass-aws/bundle/config/upstream/overlay-kube-apiserver-admission.yaml:168]: Do we test this part?

Comment (zhouke1991) [reply packages/tkg-clusterclass-aws/bundle/config/upstream/overlay-kube-apiserver-admission.yaml:168]: Tested it manually and it works well.

**Comment by zhouke1991** [comment]
test case:

1. Create mgmt cluster without APISERVER_EVENT_RATE_LIMIT_CONF_BASE64 configuration.
2. Create mgmt cluster with APISERVER_EVENT_RATE_LIMIT_CONF_BASE64 configuration but no apiserver arg "enable-admission-plugins"
3. Create mgmt cluster with APISERVER_EVENT_RATE_LIMIT_CONF_BASE64 configuration and APISERVER_EXTRA_ARGS: enable-admission-plugins=AlwaysPullImages,NodeRestriction

**Review by wjun — APPROVED** [review]
LGTM.
